fv3core/stencils/fvtp2d.py

```diff
@@ -80,16 +82,28 @@ class FiniteVolumeTransport:
     ONLY USE_SG=False compiler flag implements
     """
 
-    def __init__(self, namelist, hord, nord=None, damp_c=None):
-        self.grid = spec.grid
-        shape = self.grid.domain_shape_full(add=(1, 1, 1))
-        origin = self.grid.compute_origin()
+    def __init__(
+        self,
+        grid: GridIndexing,
+        dxa,
+        dya,
+        area,
+        grid_type: int,
+        hord,
+        nord=None,
+        damp_c=None,
+    ):
+        # TODO: add a docstring
+        self.grid = grid
+        self._area = area
+        shape = self.grid.domain_full(add=(1, 1, 1))
```

Comment:
since we are making a new object anyways, how about we add an abstraction for the shape, this is the shape used by all 3d fields in the model, could be something like domain_max, or field_shape_3d or whatever you prefer.

Reply:
Should it actually be the shape used by all 3d fields? I know it works, and that we (including me) did this because it's easy/the same everywhere and you don't have to think hard about it, but in theory all cell-centered variables should only need `domain_full` and interface variables only need an add in one or the other direction. If we did this the code would also make it clearer what's on cell centers or interfaces. I also suspect we're currently doing some computations on the last z-level that aren't needed but are never checked by the validation framework (which slices to the fortran-equivalent part of the array).

The GridIndexing object is currently written as "this provides indexing for cell-centered variables, use add if you need to use it for interface variables" which I find fairly easy to think about. I'm a little worried about adding an attribute on top of that which should only be needed for cell corner variables.

Reply:
I think a future refactor could use different shapes, but is separate from the purposes of this PR. Making it clear we are using the same shape everywhere currently with a single named variable would provide clarity. It wasn't just simpler for us porting code to use the same shape, at the time it was necessary in gt4py. That is no longer true, but we don't know if there are bugs that only pop up when you use different shapes, especially when you get regions involved. Using the same shape everywhere is also easier to understand for the user, in my opinion. I doubt there is much performance implication with doing extra computation on some buffer points. We are indeed doing some calculation on the buffer z point where we don't need to, but it does not harm as those points are never read. 
Currently this variable would be used in every single class, every 3d variable is this shape currently and is currently saying shape = grid.domain_shape_full(add=(1, 1, 1)). I think refactoring to remove global state should be a separate and distinct set of PRs from one using different sized array.s

Reply:
I'll add a `grid_indexing.max_shape` property, with a docstring explaining why it should be rarely used, to encourage us to do these refactors later.